doc/sendiso.1.in
.TH sendiso 1 "libraw1394 @VERSION@" "" "Linux IEEE 1394"
.SH NAME
sendiso \- send IEEE 1394 isochronous packets from dump file
.SH SYNOPSIS
.B sendiso
[ \fIoptions\fR ] \fIFILE\fR
.SH DESCRIPTION
sendiso takes the dump file \fIFILE\fR as generated by \fBdumpiso\fR(1),
extracts the packets and sends them as they were received (in the same order,
with the same channel numbers) as fast as possible - timing of the original dump
is not preserved.
.PP
This program does not allocate any isochronous resources but just starts
sending.  It should not be used in live environments where it might interfere
with properly set up isochronous transmissions, it is a debug and performance
benchmarking tool.
.SH OPTIONS
.TP
.B -l\fR,\fB --loop=\fICOUNT
Send dump file \fICOUNT\fR times.  Defaults to 1.
.TP
.B -i\fR,\fB --infinite
Send dump file in an endless loop.
.TP
.B -s\fR,\fB --speed=\fISPEED
Send packets at speed \fISPEED\fR, which can be given as 100, 200 and 400 (in
Mbit/s) or abbreviated as 1, 2 and 4.  Defaults to 100.
.TP
.B -p\fR,\fB --port=\fIPORT
Choose port \fIPORT\fR for sending.  A port is a 1394 card or chip and
represents one connected bus, therefore this is only relevant when you have
multiple of these.  Defaults to 0.
.TP
.B -h\fR,\fB --help
Show help text and exit.
.SH BUGS
None known.
.SH SEE ALSO
.B dumpiso\fR(1)
.SH AUTHOR
Andreas Bombe <aeb@debian.org>


NEWS
libraw1394 Release Notes

Version 2.0.4:
- Fix build due to incomplete tarball. 

Version 2.0.3:
- various bug fixes, in particular for firewire stack

Version 2.0.2:
- Fixed license headers on new files introduced in v2.0.0 to be consistent
  with LGPL v2.1.

Version 2.0.1:
- Optimized usage of /dev/fw? nodes based upon user permissions.
- many good bugfixes

Version 2.0.0:
- added support for new firewire subsystem.
- removed deprecated isochronous functions.

Version 1.3.0:
- added raw1394_read_cycle_timer() as used by FFADO project.
- added support for environment variable RAW1394DEV to override default
  device node /dev/raw1394.
- added --with-valgrind to configure script.
- other optimizations and bugfixes - see ChangeLog.

Version 1.2.1:
- generate compiler warnings on deprecated functions (legacy isochronous).
- bugfixes and optimizations.

Version 1.2.0:
- added constants useful for ARM: transaction and response return codes.
- added raw1394_bandwidth_modify() and raw1394_channel_modify().

Version 1.1.0:
- added ability to specify the isochronous receive mode in the new
  rawiso (raw1394_iso_*) API. Therefore, the signature of 
  raw1394_iso_recv_init() changed, and likewise, the libtool/.so version
  numbers changed to reflect a new interface that is not backwards compatible!
- added raw1394_arm_set_buf() and raw1394_arm_get_buf() functions.
- added raw1394_new_handle_on_port() to make initialization simpler.
- make versioning consistent and updated. 1.0.0 was reported by pkg-config in 
  0.10 series; hence, 1.1.0 in this release.
- removed outdated, bogus "linking required only" version symbols in 
  version.c - use pkg-config now.
- dropped libraw1394.m4 - use pkg-config now.
- added RPM spec file.
- deprecate old isochronous API.
- move API comment documentation to header and reformat header comments as
  kernel-doc (use linux/scripts/kernel-doc to extract and format them).
- updated tools/dumpiso and tools/sendiso to new isochronous API.

Version 0.10:

- merged in the rawiso branch
- add sendiso and dumpiso tools with man files
- fix wrong error mapping on local transactions in raw1394_errcode_to_errno().
- added comment headers for most exported functions
- merged in weihs branch which brings new functionality including address
  range mapping, reading and writing configROM images, additional bus reset
  handling, and others.

Version 0.9:

- error reporting reworked to be more C library style; functions affected are
  mainly raw1394_read(), raw1394_write(), raw1394_lock;
  Source incompatibility!  Parts need to be adapted in application code, but
  it's well worth it.
- the generation number is not unconditionally updated on bus reset in the event
  loop anymore, the new function raw1394_update_generation() has to be used
  (unless the default bus reset handler is used) so that transactions won't fail
- iso handlers used to get the iso header quadlet in host endian, now they get
  it in big endian like the rest of the packet for consistency
- testlibraw gets installed now, man page also included
- (not yet complete) documentation in Docbook format included
- libraw1394.m4 autoconf macro added

Version 0.8:

- added libraw1394/ieee1394.h with useful defines (especially ack and rcodes)
- function to query isochronous resource manager ID
- functions for isochronous sending
- new raw1394_reset_bus() function to reset the bus

Version 0.7:

- added support for environments where the kernel runs in 64 bit mode whereas
  user space runs in 32 bit mode
- fixed lock transaction to actually return the required response; these
  functions are actually useful now, prototypes for raw1394_start_lock and
  raw1394_lock changed
- a void pointer can now be stored as user data with a handle

Version 0.6:

- FCP monitoring added
- not really libraw1394 news: the subsystem doesn't swap data anymore,
  everything is to be kept big endian now on all architectures

Version 0.5:

- there are now separate iso handlers for each channel

Version 0.4:

- isochronous receive support

Version 0.3:

- first release of new libraw1394 for new raw1394 driver
- supports 1394 async reads/writes
- maintainer: Andreas Bombe <andreas.bombe@munich.netsurf.de>


tools/sendiso.c
/*
 * libraw1394 - library for raw access to the 1394 bus with the Linux subsystem.
 *
 * Copyright (C) 1999,2000 Andreas Bombe
 *
 * This library is licensed under the GNU Lesser General Public License (LGPL),
 * version 2.1 or later. See the file COPYING.LIB in the distribution for
 * details.
 */

#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <getopt.h>
#include <errno.h>
#include <fcntl.h>
#include <string.h>

#include "../src/raw1394.h"

#define BUFFER 1000
#define PACKET_MAX 4096

unsigned long which_port;
char *filename;

unsigned long loopcount = 1;
unsigned int speed;

void usage_exit(int exitcode)
{
        fprintf(stderr,
"Usage: sendiso [opts] FILE\n"
"Send IEEE 1394 isochronous packets from dump file FILE.\n"
"\n"
"-l --loop      COUNT       Repeat sending data COUNT times.\n"
"-i --infinite              Repeat sending data infinitely.\n"
"-s --speed     SPEED       Send data at SPEED (valid values are 100, 200, 400 or\n"
"                           alternatively 1, 2, 4). (default: 100)\n"
"\n"
"-p --port      PORT        Choose 1394 chip PORT. (default: 0)\n"
"-h --help                  Show this help.\n"
);

        exit(exitcode);
}

void parse_args(int argc, char **argv)
{
        char *tail;

        int c;
        int index;
        static struct option opts[] = {
                { "file",     required_argument, NULL, 'f' },
                { "loop",     required_argument, NULL, 'l' },
                { "infinite", no_argument,       NULL, 'i' },
                { "speed",    required_argument, NULL, 's' },
                { "port",     required_argument, NULL, 'p' },
                { "help",     no_argument,       NULL, 'h' },
                { 0 }
        };

        while (1) {
                c = getopt_long(argc, argv, "f:l:is:p:h", opts, &index);
                if (c == -1) break;

                switch (c) {
                case 'f':
                        filename = optarg;
                        break;
                case 'l':
                        loopcount = strtoul(optarg, &tail, 10);
                        if (*tail) {
                                fprintf(stderr,
                                        "invalid argument to loop: %s\n",
                                        optarg);
                                usage_exit(1);
                        }
                        break;
                case 'i':
                        loopcount = 0;
                        break;
                case 's':
                        speed = strtoul(optarg, &tail, 10);
                        if (*tail) speed = -1;

                        switch (speed) {
                        case 1:
                        case 100:
                                speed = RAW1394_ISO_SPEED_100;
                                break;
                        case 2:
                        case 200:
                                speed = RAW1394_ISO_SPEED_200;
                                break;
                        case 4:
                        case 400:
                                speed = RAW1394_ISO_SPEED_400;
                                break;
                        default:
                                fprintf(stderr,
                                        "invalid argument to speed: %s\n",
                                        optarg);
                                usage_exit(1);
                        }
                        break;
                case 'p':
                        which_port = strtoul(optarg, &tail, 10);
                        if (*tail) {
                                fprintf(stderr,
                                        "invalid argument to port: %s\n",
                                        optarg);
                                usage_exit(1);
                        }
                        break;
                case 'h':
                        usage_exit(0);
                case '?':
                        usage_exit(1);
                case 0:
                        break;
                default:
                        abort();
                }
        }

        argv += optind;
        argc -= optind;

        if (argc > 1) {
                fprintf(stderr, "Too many arguments.\n");
                usage_exit(1);
        }

        if (argc) filename = *argv;
}

#define BUF_SIZE 4096
#define BUF_HEAD 8
void send_file_once(raw1394handle_t handle, int file)
{
        int count, i, ret;
        unsigned channel, tag, sy;
        size_t length;
        static unsigned char buffer[BUF_SIZE + BUF_HEAD];
        static unsigned int counter = 0;
        static int inited = 0;

        while (1) {
                count = read(file, buffer, BUF_HEAD);
                if (count < 0) {
                        perror("read");
                        exit(1);
                }
                if (count < BUF_HEAD)
                        return;
        
                i = 0;
                length = ((unsigned int *)buffer)[i];
                channel = buffer[i + 4];
                tag = buffer[i + 5];
                sy = buffer[i + 6];
        
                i += BUF_HEAD;
                while (count < length + BUF_HEAD) {
                        ret = read(file, buffer + count,
                                   length - count + BUF_HEAD);

                        if (ret < 0) {
                                perror("read");
                                exit(1);

                        }
                        if (ret == 0)
                                return;

                        count += ret;
                }
                
                if (inited == 0) {
                        /*
                        fprintf(stderr, "transmitting first packet with length "
                             "%d on channel %d with tag %d and sy %d\n",
                                length, channel, tag, sy);
                        */
                        ret = raw1394_iso_xmit_init(handle, NULL, BUFFER,
                                PACKET_MAX, channel, speed, -1);
                        if (ret < 0) {
                                perror("raw1394_iso_xmit_init");
                                exit(1);
                        }
                        raw1394_iso_xmit_start(handle, -1, -1);
                        inited = 1;
                }

                if (++counter % 1000 == 0)
                        fprintf(stderr, "\r%uK packets", counter/1000);

                ret = raw1394_iso_xmit_write(handle, &buffer[i],
                        length, tag, sy);
                if (ret < 0) {
                        perror("\nraw1394_iso_xmit_write");
                        exit(1);
                }
        }
}


void send_iso_file(raw1394handle_t handle)
{
        int file;
        int count, ret;
        char buffer[32];

        if (filename[0] == '-' && filename[1] == '\0') {
                file = fileno(stdin);
        } else {
                file = open(filename, O_RDONLY, 0);

                if (file < 0) {
                        perror("open");
                        exit(1);
                }
        }

        count = 32;
        while (count) {
                ret = read(file, buffer, count);

                if (!ret) goto bad_format;

                if (ret < 0) {
                        perror("read");
                        exit(1);
                }

                count -= ret;
        }

        if (memcmp("1394 isodump v", buffer, 14)) goto bad_format;
        if (buffer[14] != '2') goto wrong_version;

        while (1) {
                send_file_once(handle, file);

                if (!loopcount) {
                        if (lseek(file, 32, SEEK_SET) < 0) {
                                perror("lseek");
                                exit(1);
                        }
                        continue;
                }
                if (!(--loopcount)) break;

                if (lseek(file, 32, SEEK_SET) < 0) {
                        perror("lseek");
                        exit(1);
                }
        }

        return;

bad_format:
        fprintf(stderr, "Input file format not recognized.\n");
        exit(1);

wrong_version:
        fprintf(stderr, "Format version of input file not supported.\n");
        exit(1);
}


int main(int argc, char **argv)
{
        raw1394handle_t handle;
	int ret;

        parse_args(argc, argv);

        fprintf(stderr, "port: %ld\nloops: %ld\nfile: %s\n", which_port,
                loopcount, filename);

        handle = raw1394_new_handle();
        if (!handle) {
                if (!errno)
                        fprintf(stderr, "No working kernel driver found.\n");
                else
                        perror("raw1394_get_handle");
                exit(1);
        }

        do {
                if (raw1394_get_port_info(handle, NULL, 0) <= which_port) {
                        fprintf(stderr, "Port %ld does not exist.\n",
                                which_port);
                        exit(1);
                }

                ret = raw1394_set_port(handle, which_port);
        } while (ret < 0 && errno == ESTALE);

        if (ret < 0) {
                perror("raw1394_set_port");
                exit(1);
        }

        if (filename)
                send_iso_file(handle);

        fprintf(stderr, "\n");
        raw1394_iso_shutdown(handle);
        raw1394_destroy_handle(handle);

        return 0;
}
